Migrate authentication services to intranet hostname, with DNS and Certbot changes

webdavis opened the pull request:
## Summary

Completes the migration of authentication services from public hostname (justdavis.com) to intranet hostname (intranet.justdavis.com), along with comprehensive fixes for related services and test infrastructure improvements.

## Problem

After restricting LDAP and Kerberos firewall rules to trusted networks, internal authentication clients could not connect because:

1. Auth clients were configured to use `justdavis.com` which resolves to the public IP of `eddings`.
2. Traffic to the public IP would require hairpin NAT.
3. Even with hairpin NAT, the source IP would appear as the router’s IP, not the client LAN IP.
4. Firewall rules block connections that do not originate from LAN (`10.0.0.0/24`) or Tailscale (`100.64.0.0/10`).
5. Authentication times out, causing failures for LDAP, Kerberos, and other dependent services.

## Major Changes

### 1. Authentication Hostname Migration

- Auth clients (SSSD, PAM/nslcd): updated to connect via `ldaps://eddings.intranet.justdavis.com`.
- Kerberos clients: updated to use `eddings.intranet.justdavis.com` as KDC.
- File server: updated Kerberos principals to include intranet hostname.
- DNS: enhanced intranet zone with proper A and NS records; updated reverse DNS.
- Test environment: mapped private IPs to intranet hostnames to support AWS VPC tests.

### 2. Let's Encrypt Certificate Fixes

- Split certificate requests between root domains and subdomains to avoid redundancy:
  - Root domains: request bare + wildcard (e.g., `justdavis.com + *.justdavis.com`).
  - Subdomains: request wildcard only (e.g., `*.intranet.justdavis.com`).
- Prevents “subdomain redundant with parent wildcard” errors.

### 3. LDAP GSSAPI Authentication Fixes

- Restored and properly configured `ldap/eddings` service principal.
- Added `ldap/intranet` service principal to handle canonicalization for intranet clients.
- Updated `olcSaslHost` to intranet hostname in `slapd-sasl.ldif`.
- Ensured proper slapd restarts when keytabs change.
- Fixed GSSAPI realm detection in test environment.

### 4. Mail Server Fixes

- Dovecot: connect to LDAP via intranet hostname (private IP).
- Mail tests: adjusted retry timing and disabled rate limiting in test environment to prevent test failures.
- Excluded intranet hostname from virtual mailbox domains to avoid duplicate handling.

### 5. File and File Server Updates

- CIFS mounts updated to use intranet hostname.
- Both `eddings` and `intranet` principals added to Samba keytab to support canonicalization.

### 6. Firewall Adjustments for Test Environment

- Added AWS VPC rules for LDAP, LDAPS, and Kerberos services in test environment.
- Production firewall rules remain restricted to LAN and Tailscale networks.

### 7. Test Infrastructure Improvements

- Teardown reliability: store EC2 instance IDs in inventory to avoid hangs.
- DNS tests: check for non-empty results instead of exact IPs for more flexibility.
- Added intranet DNS resolution tests for LDAP, Kerberos, and service subdomains.

### 8. Configuration Improvements

- Base system hostname configuration refactored to use computed prefix.
- Comprehensive comments explaining domain variable split for Let's Encrypt.
- Apache virtual host configuration updated to correctly serve intranet hostnames.

## Testing

All changes have been tested in the AWS environment.

- Authentication clients connect via intranet hostname successfully.
- LDAP GSSAPI authentication verified to work in both production and test environments.
- Mail server authentication and delivery validated.
- Test infrastructure provisions, configures, and tears down reliably.
- Let's Encrypt certificates generate without redundancy errors.

## Compatibility

- Works in production (`eddings.intranet.justdavis.com → 10.0.0.2`) and in AWS test environment: (`eddings.intranet.tests42.tests.justdavis.com → EC2 VPC private IP`).

Fixes #53
Fixes #39

webdavis committed:
refactor(intranet-dns-certbot): Completes the migration of authentication services from public hostname (`justdavis.com`) to intranet hostname (`intranet.justdavis.com`)

Problem:

  After restricting LDAP and Kerberos firewall rules to trusted
  networks, internal authentication clients could not connect because:

  1. Auth clients were configured to use `justdavis.com` which resolves
     to the public IP of `eddings`.
  2. Traffic to the public IP would require hairpin NAT.
  3. Even with hairpin NAT, the source IP would appear as the router’s
     IP, not the client LAN IP.
  4. Firewall rules block connections that do not originate from LAN
     (`10.0.0.0/24`) or Tailscale (`100.64.0.0/10`).
  5. Authentication times out, causing failures for LDAP, Kerberos, and
     other dependent services.

Major Changes:

  1. Authentication Hostname Migration:
    - Auth clients (SSSD, PAM/nslcd): updated to connect via `ldaps://eddings.intranet.justdavis.com`.
    - Kerberos clients: updated to use `eddings.intranet.justdavis.com` as KDC.
    - File server: updated Kerberos principals to include intranet hostname.
    - DNS: enhanced intranet zone with proper A and NS records; updated reverse DNS.
    - Test environment: mapped private IPs to intranet hostnames to support AWS VPC tests.

  2. Let's Encrypt Certificate Fixes:
    - Split certificate requests between root domains and subdomains to avoid redundancy:
      - Root domains: request bare + wildcard (e.g., `justdavis.com + *.justdavis.com`).
      - Subdomains: request wildcard only (e.g., `*.intranet.justdavis.com`).
    - Prevents “subdomain redundant with parent wildcard” errors.

  3. LDAP GSSAPI Authentication Fixes:
    - Restored and properly configured `ldap/eddings` service principal.
    - Added `ldap/intranet` service principal to handle canonicalization for intranet clients.
    - Updated `olcSaslHost` to intranet hostname in `slapd-sasl.ldif`.
    - Ensured proper slapd restarts when keytabs change.
    - Fixed GSSAPI realm detection in test environment.

  4. Mail Server Fixes:
    - Dovecot: connect to LDAP via intranet hostname (private IP).
    - Mail tests: adjusted retry timing and disabled rate limiting in test environment to prevent test failures.
    - Excluded intranet hostname from virtual mailbox domains to avoid duplicate handling.

  5. File and File Server Updates:
    - CIFS mounts updated to use intranet hostname.
    - Both `eddings` and `intranet` principals added to Samba keytab to support canonicalization.

  6. Firewall Adjustments for Test Environment:
    - Added AWS VPC rules for LDAP, LDAPS, and Kerberos services in test environment.
    - Production firewall rules remain restricted to LAN and Tailscale networks.

  7. Test Infrastructure Improvements:
    - Teardown reliability: store EC2 instance IDs in inventory to avoid hangs.
    - DNS tests: check for non-empty results instead of exact IPs for more flexibility.
    - Added intranet DNS resolution tests for LDAP, Kerberos, and service subdomains.

  8. Configuration Improvements:
    - Base system hostname configuration refactored to use computed prefix.
    - Comprehensive comments explaining domain variable split for Let's Encrypt.
    - Apache virtual host configuration updated to correctly serve intranet hostnames.

Testing:

  All changes have been tested in the AWS environment.

  - Authentication clients connect via intranet hostname successfully.
  - LDAP GSSAPI authentication verified to work in both production and test environments.
  - Mail server authentication and delivery validated.
  - Test infrastructure provisions, configures, and tears down reliably.
  - Let's Encrypt certificates generate without redundancy errors.

Compatibility:

  - Works in production (`eddings.intranet.justdavis.com → 10.0.0.2`) and
    in AWS test environment: (`eddings.intranet.tests42.tests.justdavis.com → EC2 VPC private IP`).

Fixes #53
Fixes #39

karlmdavis commented on roles/base/tasks/config.yml:
```diff
@@ -137,24 +137,28 @@
+- name: Compute Hostname Prefix (strip base domain)
+  ansible.builtin.set_fact:
+    hostname_prefix: "{{ inventory_hostname | regex_replace('\\.justdavis\\.com$', '') }}"
+
```
I think this is broken in tests, producing results like "`eddings.tests42.tests`" instead of just "`eddings`".

karlmdavis commented on roles/file_server/tasks/config.yml:
```diff
@@ -32,16 +32,18 @@
     - eddings
-    # Samba listens on the LAN/VPN-only IP that this resolves to.
-    - eddings.karlanderica
+    - intranet
```
Also need a restart hook attached to it.

karlmdavis replied:
(bump)

webdavis replied:
> Probably needs a restart hook attached to it.
> 
> I'm also quite confused by why the comment here and the `smb.conf.j2` file (not in this PR, but in the project already) mention a Samba keytab but don't ever appear to _create_ it? Is Samba creating its keytab automatically?

Apparently, Samba doesn't create the keytab automatically. I can't find it happening anywhere else in the playbook, either. So, this probably isn't working? 

Should I address this in a different PR?

webdavis replied:
Got it! 🫡

karlmdavis reviewed (commented):
Have a couple questions about some of the stuff in here that need some investigation. Which is great! Because the other PR was such a junk drawer, it was a **lot** harder to notice stuff like this — thanks!

webdavis commented on roles/apache/templates/justdavis.conf.j2:
```diff
@@ -1,10 +1,7 @@
-  {% endif -%}
-{% endfor -%}
+  ServerAlias www.{{ domain }}
+  ServerAlias mail.{{ domain }}
```
I'll give it a shot right now.

karlmdavis replied:
(bump)

webdavis committed:
fix(restart-slapd): ensure `slapd` reloads new service principals and SASL config

slapd only reads Kerberos credentials at startup. In cases where any of
the following happen:

  1. /etc/ldap/ldap.keytab already exists with the correct permissions
  2. A principal is added to the keytab
  3. slapd-sasl.ldif changes

slapd must be restarted so it can authenticate clients using that
new principal and apply the updated SASL settings.

webdavis committed:
refactor(mail-server-test): reduce IMAP retry frequency to avoid UFW rate limiting

The mail server test tasks were hitting the UFW rate limit (6
connections per 30 seconds). By reducing retries from 60 to 6 and
increasing the delay from 1s to 10s, the test now stays well under the
limit while still allowing multiple retries for verifying mail delivery.

webdavis committed:
fix(restart-samba): ensure Samba reloads when new service principals are added

The CIFS service principals for eddings and intranet were being created
but Samba was not restarting, meaning it would not pick up the new
Kerberos principals. By adding a notify to restart Samba after the
krb_principal tasks, the server now reloads credentials immediately,
ensuring clients can authenticate correctly.

webdavis committed:
fix(hostname-fqdn): ensure `/etc/hosts` contains the unqualified-domain-name for convenience

webdavis committed:
fix(apache-ports): remove mail subdomain server alias from Apache virtual hosts on ports 80 and 443

karlmdavis commented on roles/ldap_server/tasks/sasl-gssapi.yml:
```diff
@@ -28,18 +28,27 @@
 
+- name: Create Service Keytab for LDAP Server (intranet)
+  ansible.builtin.script: "add-keytab.sh --principal ldap/intranet.{{ domain }} --keytab /etc/ldap/ldap.keytab"
+  become: true
```
You resolved this thread because you ended up adding a restart hook?

karlmdavis reviewed (commented):
Great work: changes here all look good!

I did mark some of the earlier review threads as unresolved, though: please be sure to just leave a brief note as to why you're marking them as resolved (and in the future). Otherwise, I have to spend a bunch of time flipping between those threads and your latest commits, trying to guess at it.
